### ISSUE TITLE
GameGuru MAX - Bug - Issues with More than 5 Point Lights

### ISSUE THREAD

**Issue by ArgentArts** [TheGameCreators/GameGuruRepo#2021]
If you have more than 5 point lights in a scene, then the furthest from the camera will stop casting shadows. As you add even more point lights, others will stop casting shadows, even though they are close to the camera. 

https://user-images.githubusercontent.com/36735507/153081583-3deb29f3-c805-49db-a223-733c1c469788.mp4

**Comment by synchromesh62** [comment]
Only the nearest 4 lights can cast shadows at one time .. 
This is currently by design

**Comment by ArgentArts** [comment]
That's ... terrible. Why?

**Comment by synchromesh62** [comment]
Even the Cellar demo does it. 
Its always been that way .. If you fill a corridor with lights the 4 behind stop casting shadows and the 4 in front will start casting shadows. To overcome this they added a cast shadow tick box to lights so you could only have shadows where you really want them although not sure if its enabled yet 
![cast shadow](https://user-images.githubusercontent.com/35708217/153091598-b4eb79b6-97a7-4c28-9e22-efd492e2c3f5.png)

**Comment by UltraVox001** [comment]
@ArgentArts 
> That's ... terrible. Why? And in my video, 5 are casting shadows ... ;)

A few more shadows were added in the latest build. But obviously that's not enough.

**Comment by synchromesh62** [comment]
**And in my video, 5 are casting shadows ... ;)**
That may be a bug and need putting back to 4.
Just kidding :)
Im going by the info Preben gave me when i asked him about the very same issue with my first Max Demo so maybe its been upped to 5 lights since then ?

**Comment by UltraVox001** [comment]
The answer is here : [https://github.com/TheGameCreators/GameGuruRepo/issues/1830](https://github.com/TheGameCreators/GameGuruRepo/issues/1830)

**Comment by synchromesh62** [comment]
@UltraVox001 
Not the same issue. 
We are referring to the amount of lights that can cast shadows at any one time.

**Comment by synchromesh62** [comment]
> The answer is here : #1830

Not the same issue.
We are referring to the amount of lights that can cast shadows at any one time.

**Comment by LeeBamberTGC** [comment]
We limit the POINT LIGHT shadows to process the nearest FIVE as they should be used deliberately and with care due to their performance hit (they render the scene SIX times per point light).  Here is a typical shot of point and spot lights:

![image](https://user-images.githubusercontent.com/35840583/153199638-e24bd561-9e96-4650-960e-55a32da007c8.png)

**Comment by LeeBamberTGC** [comment]
@ArgentArts Shadows rendered for the "nearest FIVE", you can have hundreds over the course of your level.  I am currently fixing a bug so the cast shadow flag in the light properties works, before it was not connected up, thus not very useful for level design.

**Comment by ArgentArts** [comment]
I know that it renders the nearest five. But there are instances where a level designer needs more than 5 ... a lot more than five. This is why most engines provide both STATIC and DYNAMIC lighting, so you can get as many static shadows as you require to bring your level to life, casting shadows from every little light source.

**Comment by LeeBamberTGC** [comment]
When you stated: "it should be up to the person(s) creating their game how many lights casting shadows they want to put in each of their levels" I assumed you referred to lights that can cast shadows for the whole level.  Users can have every light they put down, cast a shadow, in fact this is the default. Most games that rely entirely on dynamic real-time lighting manage this resource carefully for performance considerations, and any solution needs to consider a level designed on a high-end rig running on a low-end rig.  I think it will be a few more years before game developers can throw away the consideration of performance when lighting their scenes.

Perhaps you can send me a scene that needs more than five nearby lights, all casting shadows, and I will have a think about how this can be achieved in real-time (perhaps as a high-end only option).  There are no current plans to use lightmapping for MAX but Wicked Engine does support a decent if basic light mapper so there might be options too for the lower end users down the road.

**Comment by UltraVox001** [comment]
Personally, I am constantly confronted with this problem of light-shadows. These are not small 3D levels that are easy to make or play, like most simplistic demos are. These are scenes with very complex lights and shadows. The slightest step forward, or backward, makes all the shadows disappear in a very important visual area. It is so blatant that it poses enormous problems. 

Maybe we should allow each light to cast shadows differently depending on the player's graphics settings : Low, Medium, High, and finally, **Ultra**.

Ultra graphics would allow the lights to be set differently than with lower settings intended for smaller machines.
It's just a suggestion :)

**Comment by UltraVox001** [comment]
Video games like Icarus (new Survival game on Steam) or Star Citizen, don't care about small machines. It is the choice of the developers of these games to be able to play only if you have a PC that runs on nuclear power. :) But it's an assumed choice.

**Comment by LeeBamberTGC** [comment]
@ArgentArts @UltraVox001 Let's look at this again after March, and see if anyone else struggles to create their dynamic scenes.

**Comment by LeeBamberTGC** [comment]
I will bump the max count to 16 for this weeks build to see if that gives greater wiggle room whilst keeping performance and memory in check (we have been doing work to reduce memory use in the shadow system).

**Comment by ArgentArts** [comment]
16 sounds very reasonable. I will run some tests to see how it affects frame rates on my system and report them.

**Comment by synchromesh62** [comment]
> Thank you! :) 16 sounds very reasonable. I will run some tests to see how it affects frame rates on my system and report them.

16 is more than Leadwerk's so that's awesome plus if Lee enables the tickbox in light preferences for light entities to cast shadows or not that will be a game changer :)

**Comment by LeeBamberTGC** [comment]
@synchromesh62 Consider it done.  Will be in this evenings internal build :)

**Comment by ArgentArts** [comment]
Looking forward to testing it this Friday. :)

**Comment by UltraVox001** [comment]
As I wrote in another post :
Since Lee's last changes regarding the Cast Shadow option of lights and the increase up to 16 dynamic lights, some maps were running with 35-45 FPS, now render at 60 FPS constant. Tested on an old i5 with a GTX 1060 (6 Gigas), and 16 Gigas Ram. I congratulate everyone on this spectacular step forward. :)
It's a military base in which there are more than 400 3D objects all around me :
![60FPS](https://user-images.githubusercontent.com/96718175/153742896-033cca9e-94e8-4d05-821d-2615e2fbfb08.jpg)
